### DISCUSSION TITLE
How to add extra fonts to fitz

### DISCUSSION THREAD

**Discussion by LastMonody** [pymupdf/PyMuPDF#1498] [Looking for help]
Excuse me, if I want to get whether the font of the page is bold, but the font of my page page is not the basic 14 types in fitz, is there a way to make fitz support my font in this situation?

**Comment by JorjMcKie** [comment]
You mean you want to write on some page using a font of your choice - which is not one of the Base 14 fonts?
No problem, if your font is not too exotic.
Look in the documentation for `page.insert_font()` or about the `Font` and `TextWriter` classes.

**Reply by LastMonody** [reply]
Yes, but my current purpose is not to write text like pdf, but just to analyze the attributes of the font to determine whether it is bold or italicized.

**Reply by JorjMcKie** [reply]
Ok. Look at `page.get_fonts()`. Then extract the font in question using its xref and make a font object from the binary.
That font object can then be checked for properties:
```python
>>> from pprint import pprint
>>> import fitz
>>> doc=fitz.open("pymupdf.pdf")
>>> page=doc[0]
>>> pprint(page.get_fonts())
[(1022, 'pfa', 'Type1', 'FVHEAV+NimbusSanL-Bold', 'F49', ''),
 (1023, 'pfa', 'Type1', 'ZCYZLD+NimbusSanL-BoldItal', 'F50', '')]
>>> f = doc.extract_font(1023)  # for example
>>> font = fitz.Font(fontbuffer=f[-1])
>>> font.name
'Nimbus Sans L Bold Italic'
>>> font.is_bold
True
>>> font.is_italic
True
>>> # or do this:
>>> pprint(font.flags)
{'bold': 1,
 'fake-bold': 0,
 'fake-italic': 0,
 'invalid-bbox': 0,
 'italic': 1,
 'mono': 0,
 'opentype': 0,
 'serif': 1,
 'stretch': 0,
 'substitute': 0}
>>> 
```
Things can be more complicated in PDF - as usual:
There are tricks that let a font **_appear_** bold, although it is normal weight - just write each character twice with a small horizontal offset, or use PDF's text rendering mode 2, etc.
But the above is a good start ... if the font is supported by MuPDF at all.

**Reply by JorjMcKie** [reply]
Similar approach if you want to analyse the font file (and not have a PDF like above): `font = fitz.Font(fontfile="...")`.

**Reply by LastMonody** [reply]
But I will encounter fonts with such values (440,'n/a','TrueType','Times New Roman,Italic','F7','WinAnsiEncoding'), so an exception will be thrown when creating the object.

**Reply by LastMonody** [reply]
For example, there is still Chinese in my pdf, and the name of the Chinese font displayed will be garbled.

**Reply by JorjMcKie** [reply]
Not extractable fonts ("n/a") cannot be investigated like this, sorry.

**Reply by LastMonody** [reply]
Okay, thank you for your reply!
